### PR TITLE
DT-1614 show group code when deleting and add error handling

### PR DESCRIPTION
**Pull request by karenmillermoj**

### CODE REVIEW

File: backend/controllers/groupDelete.js [hunk 1]
```diff
@@ -1,23 +1,33 @@
 const groupDeleteFactory = (getGroupDetailsApi, deleteGroupApi, maintainUrl) => {
-  const stashStateAndRedirectToIndex = (req, res, errors, group) => {
+  const stashStateAndRedirectToIndex = (req, res, errors, group, url) => {
```
Comment (petergphillips) [inline backend/controllers/groupDelete.js:2]: can't see this extra parameter used anywhere?

File: backend/controllers/getGroups.test.js [hunk 2]
```diff
@@ -8,8 +8,9 @@ describe('select groups factory', () => {
     it('should call groups render', async () => {
       getGroups.mockResolvedValue([{ groupName: 'name', groupCode: 'code' }])
 
+      const req = { flash: jest.fn() }
```
Comment (petergphillips) [inline backend/controllers/getGroups.test.js:11]: would be good to test that `flash` is called by the `render`

File: backend/controllers/groupDelete.js [hunk 3]
```diff
@@ -1,23 +1,33 @@
...
+      res.render('groupDelete.njk', {
+        group,
+        groupUrl,
+        groupDetails,
+        hasMaintainAuthUsers,
+        maintainUrl,
+      })
+    } catch (error) {
+      if (error.status === 404) {
+        const groupChildError = [{ href: '#groupCode', text: 'Group does not exist' }]
```
Comment (petergphillips) [inline backend/controllers/groupDelete.js:24]: variable named incorrectly?  should be `groupError`?

File: backend/controllers/groupDetails.js [hunk 4]
```diff
@@ -8,14 +8,25 @@ const groupDetailsFactory = (getGroupDetailsApi, deleteChildGroupApi, maintainUr
...
+
+      res.render('groupDetails.njk', {
+        groupDetails,
+        hasMaintainAuthUsers,
+        maintainUrl,
+        errors: req.flash('deleteGroupErrors'),
+      })
+    } catch (error) {
+      if (error.status === 404) {
+        const groupChildError = [{ href: '#groupCode', text: 'Group does not exist' }]
```
Comment (petergphillips) [inline backend/controllers/groupDetails.js:23]: again should be `groupError`

File: backend/controllers/groupDelete.test.js [hunk 5]
```diff
@@ -37,5 +37,20 @@ describe('group delete factory', () => {
...
+      error.status = 404
+      // @ts-ignore
+      error.response = { body: { error_description: 'not valid' } }
+
+      const req = { params: { group: 'DOES_NOT_EXIST' }, flash: jest.fn() }
+      const render = getGroupDetailsApi.mockRejectedValue(error)
+      const redirect = jest.fn()
+
+      await groupDelete.deleteGroup(req, { redirect, render })
+      expect(redirect).toBeCalledWith('/manage-groups')
```
Comment (petergphillips) [inline backend/controllers/groupDelete.test.js:53]: could also test that `flash` is called

File: backend/controllers/groupDetails.test.js [hunk 6]
```diff
@@ -24,14 +24,29 @@ describe('Group details factory', () => {
...
+      error.status = 404
+      // @ts-ignore
+      error.response = { body: { error_description: 'not valid' } }
+
+      const req = { params: { group: 'DOES_NOT_EXIST' }, flash: jest.fn() }
+      const render = getGroupDetailsApi.mockRejectedValue(error)
+      const redirect = jest.fn()
+
+      await groupDetails.index(req, { redirect, render })
+      expect(redirect).toBeCalledWith('/manage-groups')
```
Comment (petergphillips) [inline backend/controllers/groupDetails.test.js:48]: could also test that `flash` is called

File: backend/middleware/errorHandler.js [hunk 7]
```diff
@@ -2,12 +2,17 @@ const { serviceUnavailableMessage } = require('../common-messages')
 const { logError } = require('../logError')
 
 const errorHandler = (error, req, res, next) => {
+  if (error.response.status === 404) {
+    res.status(404)
```
Comment (petergphillips) [inline backend/middleware/errorHandler.js:6]: would be good to test this scenario too

File: integration-tests/pages/errorPage.js [hunk 8]
```diff
@@ -1,6 +1,6 @@
 const page = require('./page')
 
-const errorPage = () => page('Sorry, there is a problem with the service', {})
+const errorPage = () => page('Page not found', {})
```
Comment (petergphillips) [inline integration-tests/pages/errorPage.js:3]: this looks like a strange change - don't we need a `notFoundPage.js` too?

File: backend/middleware/errorHandler.js [hunk 9]
```diff
@@ -2,12 +2,17 @@ const { serviceUnavailableMessage } = require('../common-messages')
...
+    return res.render('notFound.njk')
```
Comment (petergphillips) [inline backend/middleware/errorHandler.js:7]: think this should also include the `originalUrl` in the render so that `Continue` button works.  Might be better to do this in a separate PR as should really have an int test for it too

File: backend/controllers/groupDetails.js (hunk 4 above)
Comment (karenmillermoj) [reply backend/controllers/groupDetails.js:23]: Updated (cut and paste error)

File: backend/controllers/groupDelete.js (hunk 3 above)
Comment (karenmillermoj) [reply backend/controllers/groupDelete.js:24]: Updated (cut and paste error)

File: backend/controllers/getGroups.test.js (hunk 2 above)
Comment (karenmillermoj) [reply backend/controllers/getGroups.test.js:11]: added

File: backend/controllers/groupDelete.test.js (hunk 5 above)
Comment (karenmillermoj) [reply backend/controllers/groupDelete.test.js:53]: added

File: backend/controllers/groupDetails.test.js (hunk 6 above)
Comment (karenmillermoj) [reply backend/controllers/groupDetails.test.js:48]: added

File: backend/middleware/errorHandler.js (hunk 9 above)
Comment (karenmillermoj) [reply backend/middleware/errorHandler.js:7]: yes - will pull new error handling to separate PR

File: backend/middleware/errorHandler.js (hunk 7 above)
Comment (karenmillermoj) [reply backend/middleware/errorHandler.js:6]: yes - will pull new error handling to separate PR and add in

File: integration-tests/pages/errorPage.js (hunk 8 above)
Comment (karenmillermoj) [reply integration-tests/pages/errorPage.js:3]: will check and confirm - will pull new error handling to separate PR

File: backend/controllers/groupDelete.js (hunk 1 above)
Comment (karenmillermoj) [reply backend/controllers/groupDelete.js:2]: It is correct.  The call to stashStateAndRedirectToIndex needed an update (if attempt to manipulate URL to delete a group with child groups e.g. /manage-groups/{Group with Children}/delete/children/none).  Fixed.

**Review by petergphillips — APPROVED** [review]
LGTM
